src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -0,0 +1,292 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Helper type for common session case
+    /// 
+    /// This will tie <see cref="Solution"/> and <see cref="RemoteHostClient.Session"/>'s lifetime together
+    /// so that one can handle those more easily
+    /// </summary>
+    internal sealed class SolutionAndSessionHolder : IDisposable
```

Comment:
I am not sure why you are keep getting back to name Session. so do you want me to rename something that is called Session right now to different name and rename this to Session? like you said, caller doesn't need to call what you get back, it has all the convenience they need. IDispose that let it to be used in usings and all Invoke they can use to call the other side.